lib/pengine/complex.c

```diff
@@ -435,6 +435,60 @@ detect_promotable(pe_resource_t *rsc)
     return FALSE;
 }
 
+static void
+free_params_table(gpointer data)
+{
+    g_hash_table_destroy((GHashTable *) data);
+}
+
+/*!
+ * \brief Get a table of resource parameters
+ *
+ * \param[in] rsc       Resource to query
+ * \param[in] node      Node for evaluating rules (NULL for defaults)
+ * \param[in] data_set  Cluster working set
+ *
+ * \return Hash table containing resource parameter names and values
+ *         (or NULL if \p rsc or \p data_set is NULL)
+ * \note The returned table will be destroyed when the resource is freed, so
+ *       callers should not destroy it.
+ */
+GHashTable *
+pe_rsc_params(pe_resource_t *rsc, pe_node_t *node, pe_working_set_t *data_set)
+{
+    GHashTable *params = NULL;
+
+    /* A NULL node is used to request the resource's default parameters
+     * (not evaluated for node), but we always want something non-NULL
+     * as a hash table key.
+     */
+    const char *node_name = "";
```

Comment:
It is probably safe but somehow the assumption that this literal assignment will always be the same pointer leaves a strange feeling.
Don't know if that is more safe:
static const char *node_empty = "";
const char *node_name = node_empty;

Reply:
It's not relying on the pointer -- it's strdup()'d for the table, and the table uses the usual comparator that does a string comparison.

Reply:
sorry yeah we are using crm_strcase_hash.
we could actually for performance reasons use the pointers directly and so I was somehow assuming we did.
i guess at least it would work as we are not passing the node-name as a string but the node-pointer - we shouldn't have copies of the nodes so that the uname pointer would be different - right.

Reply:
I suspect we do need to use a string comparison, the scheduler sometimes makes copies of the node objects.

Reply:
It's a shallow copy though, node->details doesn't get copied, so actually it should be possible to use the uname pointer, or even the details pointer

Reply:
I'll leave that for a separate PR though, it will need a lot of testing

Reply:
Do you have a test-case for profiling?
So that we could quickly find out if it is worth the risk & test-effort.
Should really just be replacing of the hash-function and the compare-function.
For the key-free-function NULL is accepted.

Reply:
I just ran through the scheduler regression tests with pcmk_simtimes, plus I wrote a similar script to track memory usage (which I should probably merge into pcmk_simtimes but haven't gotten around to it)